File: Design-Patterns-Using-CSharp/Template/TemplatePattern.cs
using System;

abstract class AbstractClass
{
    // The template method
    public void TemplateMethod()
    {
        Operation1();
        Operation2();
    }

    // Primitive operation 1
    protected abstract void Operation1();

    // Primitive operation 2
    protected abstract void Operation2();
}

class ConcreteClass1 : AbstractClass
{
    protected override void Operation1()
    {
        Console.WriteLine("Concrete Class 1 : Operation 1");
    }

    protected override void Operation2()
    {
        Console.WriteLine("Concrete Class 1 : Operation 2");
    }
}

class ConcreteClass2 : AbstractClass
{
    protected override void Operation1()
    {
        Console.WriteLine("Concrete Class 2 : Operation 1");
    }

    protected override void Operation2()
    {
        Console.WriteLine("Concrete Class 2 : Operation 2");
    }
}

// Client code
class TemplatePattern
{
    public static void Main(string[] args)
    {
        AbstractClass concreteClass = new ConcreteClass1();
        concreteClass.TemplateMethod();
    }
}

File: Design-Patterns-Using-Java/Template/TemplatePattern.java
abstract class AbstractClass {
    final void templateMethod() {
        operation1();
        operation2();
    }

    abstract void operation1();
    abstract void operation2();
}

class ConcreteClass1 extends AbstractClass {
    @Override
    void operation1() {
        System.out.println("Concrete Class 1 : Operation 1");
    }

    @Override
    void operation2() {
        System.out.println("Concrete Class 1 : Operation 2");
    }
}

class ConcreteClass2 extends AbstractClass {
    @Override
    void operation1() {
        System.out.println("Concrete Class 2 : Operation 1");
    }

    @Override
    void operation2() {
        System.out.println("Concrete Class 2 : Operation 2");
    }
}

public class TemplatePattern {
    public static void main(String[] args) {
        AbstractClass concreteClass = new ConcreteClass1();
        concreteClass.templateMethod();
    }
}

/*
Concrete Class 1 : Operation 1
Concrete Class 1 : Operation 2
*/